lightning/src/chain/channelmonitor.rs

```diff
@@ -2591,6 +2634,180 @@ impl<Signer: WriteableEcdsaChannelSigner> ChannelMonitorImpl<Signer> {
+					let trusted_commitment_tx = commitment_tx.trust();
+					let commitment_txid = trusted_commitment_tx.txid();
+					let commitment_outputs = &trusted_commitment_tx.built_transaction().transaction.output;
+					let output_idx = self.find_revokeable_output(&commitment_outputs, &their_per_commitment_point)?;
+					let value = commitment_outputs[output_idx].value;
+
+					Some(RevokeableOutputData {
+						commitment_number: *commitment_number,
+						commitment_txid,
+						output_idx,
+						value,
+					})
+				},
+				_ => None,
+			}
+		}).collect()
+	}
+
+	fn revokeable_redeemscript(&self, their_per_commitment_point: &PublicKey) -> Script {
+		let revocation_pubkey = chan_utils::derive_public_revocation_key(&self.onchain_tx_handler.secp_ctx, &their_per_commitment_point, &self.holder_revocation_basepoint);
+		let delayed_key = chan_utils::derive_public_key(&self.onchain_tx_handler.secp_ctx, &their_per_commitment_point, &self.counterparty_commitment_params.counterparty_delayed_payment_base_key);
+		chan_utils::get_revokeable_redeemscript(&revocation_pubkey, self.counterparty_commitment_params.on_counterparty_tx_csv, &delayed_key)
+	}
+
+	fn find_revokeable_output(&self, counterparty_commitment_outputs: &[TxOut], their_per_commitment_point: &PublicKey) -> Option<usize> {
+		let revokeable_redeemscript = self.revokeable_redeemscript(&their_per_commitment_point);
+		let revokeable_p2wsh = revokeable_redeemscript.to_v0_p2wsh();
+		counterparty_commitment_outputs.iter().enumerate().find(|(_, out)| out.script_pubkey == revokeable_p2wsh).map(|(idx, _)| idx)
+	}
+
+	pub(crate) fn build_justice_tx(&self, counterparty_commitment_txid: Txid, vout: u32, value: u64) -> Transaction {
+		let input = vec![TxIn {
+				previous_output: bitcoin::OutPoint {
+					txid: counterparty_commitment_txid,
+					vout,
+				},
+				script_sig: Script::new(),
+				sequence: Sequence::ENABLE_RBF_NO_LOCKTIME,
+				witness: Witness::new(),
+		}];
+		let output = vec![TxOut {
+			script_pubkey: self.destination_script.clone(),
+			value,
+		}];
+		Transaction {
+			version: 2,
+			lock_time: bitcoin::PackedLockTime::ZERO,
+			input,
+			output,
+		}
+	}
+
+	pub(crate) fn sign_justice_tx(&self, mut justice_tx: Transaction, input_idx: usize, value: u64, commitment_number: u64) -> Result<Transaction, ()> {
+		let secret = self.get_secret(commitment_number).ok_or(())?;
+		let per_commitment_key = SecretKey::from_slice(&secret).map_err(|_| ())?;
+		let their_per_commitment_point = PublicKey::from_secret_key(&self.onchain_tx_handler.secp_ctx, &per_commitment_key);
+		let revokeable_redeemscript = self.revokeable_redeemscript(&their_per_commitment_point);
+
+		let sig = self.onchain_tx_handler.signer.sign_justice_revoked_output(&justice_tx, input_idx, value, &per_commitment_key, &self.onchain_tx_handler.secp_ctx)?;
```

Comment:
Between the two approaches mentioned “Get counterparty commitment txs from monitor update” and “Get revoked output data from update, build/sign separately”, I think one of the determinant factor to pick up which one might be the most advantageous is the flexibility preserved for the watchtower.

Namely, they are few characteristics than a watchtower signing policy can enforce:
- the amount of fees paid by the `justice_tx`
- the destination `scriptPubkey`
- the claim aggregation of one or more `RevokeableOutput`

In term of a potential watchtower signing policy, from my understanding the two approaches are equivalent:
- with approach 1, as we’re calling into `OnchainTxHandler`’s signer, a signing policy can be implemented behind `EcdsaChannelSigner::sign_justice_revoked_output`
- with approach 2, `build_justice_tx` is a pub method, the caller can verify the checks on the `Transaction` and then give it back to `sign_justice_tx`.

As we have already refactored our signing interface to let custom signing policy be implemented there, this is unclear to me what are the other trade-off between both approaches ? Do we wanna the transaction broadcast to be managed by the own transaction-relay module of the watchtower ? Or rather extend `OnchainTxHandler` to have `justice_tx` re-scheduled by the watchtower.

Reply:
> (in case in the future a watchtower module wanted to do anything beyond the most basic justice tx, e.g. add an output to the justice tx like mentioned [here](https://github.com/sr-gi/bolt13/blob/master/13-watchtowers.md#payment-modes) or claiming revoked HTLC outputs).

Yes adding an output as a “service bounty” for the outsourced watchtower makes a lot of sense (or fee-bumping purpose if you wanna do CPFP). Note, I would favor to expose the full commitment transaction data that way the watchtower can build claim on the revoked HTLC outputs, without waiting counterparty confirmation of second-stage HTLC transaction (which might never happen!). The anchor output could be swept too by the watchtower as recommended by BOLT3 after the CSV 16: https://github.com/lightning/bolts/blob/master/03-transactions.md#to_local_anchor-and-to_remote_anchor-output-option_anchors (not punishment per se though as you have the infrastructure laid out).

> As I've had the basic "outsourced" watchtower in mind for this, I imagined that this would expose the necessary information to build, sign, and encrypt the justice tx, and send it to the watchtower server which would take care of broadcasting the transaction separately.

“Outsourced” watchtower sounds good. There is one design option moving forward, making `OnchainTxHandler` an interface to which transaction issuer (watchtower, dual-funding rbfing) can subscribe broadcasting transaction. A robust transaction rebroadcast and fee-bumping module is easier said than done :) (especially with all transaction relay policy maintenance).